### ISSUE TITLE
make it easier to find out how a var was `set`

### ISSUE THREAD

**Issue by ghost** [fish-shell/fish-shell#1226]
Hello,

I ran into a problem when my home folder was moved and the default environment variable values in fishd.<host> were still pointing to the old directories.
I wasn't aware that the file overwrote the defaults and by calling `fish --debug-level=5` I still wasn't able to ascertain if the default values had been hard-coded, compiled using #ifdefs , etc.

Only a grep of the faulty values in all files with "fish" in their name (`grep yada $(find -name "*fish*" ~)`) gave me a hint and `rm ~/.config/fish/fishd.*` gave me resolve.

**TL;DR Please adding logging to the cpp-function called when calling `set` in fish.**

Cheers!

**Comment by krader1961** [comment]
There are two things we should do here. First, improve the debug messages so that when a var is created, modified, erased that action is logged at a debug level (probably two or three since this is fairly fundamental information). Second, add state to each var to track where it was modified. Then add an option to the `set` command (perhaps `set -qv $varname`) to display that state. This is similar to Vim's ability to ask where an option was last set; e.g., `:verbose set expandtab?` might output

```
  expandtab
        Last set from /usr/local/Cellar/vim/8.0.0084/share/vim/vim80/indent.vim
```

**Comment by krader1961** [comment]
Note that as of today if you build from the git `master` or `major` branches you no longer need to do the series of `set -q` statements in [@zanchey's comment](https://github.com/fish-shell/fish-shell/issues/1226#issuecomment-32149175) to figure out which scopes a variable appears in. You can simply do `set --show varname`. Now that we have that feature I would love to track where the var was last set and include that in the `set --show` output.
